src/main/java/uk/gov/cabinetoffice/csl/service/csrs/OrganisationalUnitService.java

```diff
@@ -93,28 +99,109 @@ public Map<Long, List<OrganisationalUnit>> getHierarchies(List<Long> organisatio
+        OrganisationalUnitMap organisationalUnitMap = getOrganisationalUnitMap();
+        OrganisationalUnit organisationalUnit = organisationalUnitMap.get(organisationalUnitId);
+        if (organisationalUnit == null) {
+            log.error("OrganisationalUnit not found for id: {}", organisationalUnitId);
+            throw new NotFoundException("OrganisationalUnit not found for id: " + organisationalUnitId);
+        }
+
+        Long originalParentId = organisationalUnit.getParentId();
+        if(originalParentId != null) {
+            OrganisationalUnit originalParent = organisationalUnitMap.get(originalParentId);
+            originalParent.getChildIds().remove(organisationalUnitId);
+        }
+
+        OrganisationalUnit newParent = parseParent(organisationalUnitDto.getParent(), organisationalUnitMap);
+        organisationalUnit.setParent(newParent);
+        organisationalUnit.setParentId(newParent != null ? newParent.getId() : null);
+        organisationalUnit.setAbbreviation(organisationalUnitDto.getAbbreviation());
+        organisationalUnit.setCode(organisationalUnitDto.getCode());
+        organisationalUnit.setName(organisationalUnitDto.getName());
+
+        List<OrganisationalUnit> multipleOrgs = getOrganisationsWithChildrenAsFlatList(singletonList(organisationalUnitId));
+        OrganisationalUnitMap updatedOrganisationalUnitMap = updateFormattedName(multipleOrgs, organisationalUnitMap);
+        log.info("Cache is updated for the organisational unit and its children: {}", organisationalUnitMap.get(organisationalUnitId));
+        organisationalUnitMapCache.put(updatedOrganisationalUnitMap);
+        return organisationalUnitMap;
+    }
+
+    private OrganisationalUnitMap updateFormattedName(List<OrganisationalUnit> organisationalUnits, OrganisationalUnitMap organisationalUnitMap) {
+        organisationalUnits
+            .forEach(o -> {
+                StringBuilder formattedName = new StringBuilder(o.getNameWithAbbreviation());
+                Long parentId = o.getParentId();
+                int parents = 0;
+                while (parentId != null) {
+                    OrganisationalUnit parentOrganisationalUnit = organisationalUnitMap.get(parentId);
+                    if (parents == 0) {
+                        parentOrganisationalUnit.addChildId(o.getId());
+                        parents++;
+                    }
+                    formattedName.insert(0, parentOrganisationalUnit.getNameWithAbbreviation() + " | ");
+                    parentId = parentOrganisationalUnit.getParentId();
+                }
+                o.setFormattedName(formattedName.toString());
+                organisationalUnitMap.put(o.getId(), o);
+            });
+        return organisationalUnitMap;
+    }
+
+    private OrganisationalUnit parseParent(String parentStr, OrganisationalUnitMap map) {
```

Comment:
There's code to strip the ID from the parentStr in the frontend, can we pass this into the endpoint as a Long please?